### PR TITLE
Implement interpolation for quadrature elements

### LINKED ISSUE #1546: Interpolation into Quadrature (non-basix) element
Elements which are not registered in basix, but still interpolatory, could prepare interpolation points and interpolation matrix during element construction in `FiniteElement.cpp`.
For Quadrature element this requires quadrature scheme provided from FFCx generated code, or better, a "variant".

### PR DESCRIPTION
**Pull request by mscroggs**
Requires https://github.com/FEniCS/ffcx/pull/634.

Resolves #1546.

Progress towards #2872

### CODE REVIEW

**Comment by chrisrichardson** [comment]
I am a bit confused by the complexity of it - it seems to be adding more generated code (which we generally try to avoid) to pass data from basix via ffcx back to dolfinx. Is this avoidable?

**Comment by mscroggs** [comment]
> I am a bit confused by the complexity of it - it seems to be adding more generated code (which we generally try to avoid) to pass data from basix via ffcx back to dolfinx. Is this avoidable?

Writing the quadrature rules into the generated code is definitely necessary.

We could avoid writing the map type into generated code if we make the assumption that elements with custom quadrature always use an identity pullback. This feels dangerous though.

The key issue is that quadrature elements are only defined in Python, so we need to write what we need into the generated code so that C++ can get hold of it

**Comment by jorgensd** [comment]
> We could avoid writing the map type into generated code if we make the assumption that elements with custom quadrature always use an identity pullback. This feels dangerous though.

I'm not sure it is dangerous. QuadratureElements isn't really a finite element (as discussed earlier), and is a way of representing data at points. Having a pull back != identity on point data seems strange.

**Comment by mscroggs** [comment]
> > We could avoid writing the map type into generated code if we make the assumption that elements with custom quadrature always use an identity pullback. This feels dangerous though.
> 
> I'm not sure it is dangerous. QuadratureElements isn't really a finite element (as discussed earlier), and is a way of representing data at points. Having a pull back != identity on point data seems strange.

The danger I'm worried about it that we're not checking if it's a quadrature element, we're checking if there's a custom quadrature rule. If at some point in the future we decided to implement (eg) a Nedelec-style element that has a custom quadrature rule (for whatever reason), this would lead to odd behaviour.

**Comment by mscroggs** [comment]
> > > We could avoid writing the map type into generated code if we make the assumption that elements with custom quadrature always use an identity pullback. This feels dangerous though.
> > 
> > 
> > I'm not sure it is dangerous. QuadratureElements isn't really a finite element (as discussed earlier), and is a way of representing data at points. Having a pull back != identity on point data seems strange.
> 
> The danger I'm worried about it that we're not checking if it's a quadrature element, we're checking if there's a custom quadrature rule. If at some point in the future we decided to implement (eg) a Nedelec-style element that has a custom quadrature rule (for whatever reason), this would lead to odd behaviour.

I've changed it to check whether the element is a quadrature element directly, so no longer any need to worry about possible other future uses of custom quadrature

File: cpp/dolfinx/fem/FiniteElement.h [hunk 1]
```diff
@@ -690,5 +690,9 @@ class FiniteElement
 
   // Basix Element (nullptr for mixed elements)
   std::unique_ptr<basix::FiniteElement<T>> _element;
+
+  // Quadrature points of a quadrature element (nullptr for non-quadrature
```
Comment (garth-wells) [inline cpp/dolfinx/fem/FiniteElement.h:694]: Not a pointer, so can't be `nullptr`.

**Comment by jorgensd** [comment]
Could we merge this?

**Pull request merged**